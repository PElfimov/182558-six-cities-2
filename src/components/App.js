import React from "react";
import Main from "./mainPage/Main/Main";

export default function App() {
  return (
    <Main />
  );
}

