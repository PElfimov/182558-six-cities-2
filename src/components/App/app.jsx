import React, {PureComponent} from "react";
import {connect} from "react-redux";
import Header from "../header/header";
import City from "../main-page/city/city";
import Tabs from "../tabs-panel/tabs/tabs";
import propTypes from "./prop-types";
import {ActionCreator, getCitiesListFromOffers, getFilteredOffers} from "../../store/actions/action-creator/action-creator";
import withActiveCard from "../../hocs/with-active-card/with-active-card";
import ModelOffers from '../../store/model-offers/model-offers';
import SignIn from '../sign-in/sign-in';
import withSignIn from '../../hocs/with-sign-in/with-sign-in';
import {Switch, Route, Redirect} from "react-router-dom";


const WithActiveCard = withActiveCard(City);
const SignInWrapped = withSignIn(SignIn);

class App extends PureComponent {
  constructor(props) {
    super(props);
  }

  _initialState() {
    const cities = getCitiesListFromOffers(this.props.offers);
    const offers = getFilteredOffers(this.props.offers, cities[0]);
    this.props.setCities(cities);
    this.props.changeCity(cities[0], offers);

  }

  replaceOffers(city) {
    const {offers} = this.props;
    const citiesOffers = getFilteredOffers(offers, city);
    this.props.changeCity(city, citiesOffers);
  }
  componentDidUpdate(prevProps) {
    if (this.props.offers.length !== prevProps.offers.length) {
      this._initialState();
    }
  }

  renderScreen(isAuthorizationRequired) {
    if (isAuthorizationRequired) {
      const {city, cities, cityOffers} = this.props;
      return <div className="page page--gray page--main">
        <Header />
        <main className="page__main page__main--index">
          <h1 className="visually-hidden">Cities</h1>
          <Tabs
            cities={cities}
            activeCity={city}
            onChangeCity={(selectedCity) => this.replaceOffers(selectedCity)}
          />
          <WithActiveCard offers={cityOffers} />
        </main>
      </div>;
    } else {
      return <SignInWrapped />;
    }
  }


  render() {
    const {isAuthorizationRequired} = this.props;

    return (
      <React.Fragment>
        {this.renderScreen(isAuthorizationRequired)}
      </React.Fragment>
    );
  }
}

App.propTypes = propTypes;

const mapStateToProps = (state, ownProps) =>
  Object.assign({}, ownProps, {
    city: state.localData.city,
    cityOffers: state.localData.cityOffers,
    cities: state.localData.cities,
    offers: ModelOffers.parseOffers(state.externalData.offers),
    isAuthorizationRequired: state.externalData.isAuthorizationRequired,

  });

const mapDispatchToProps = (dispatch) => ({
  setCities: (cities) => {
    dispatch(ActionCreator.setCities(cities));
  },
  changeCity: (city, offers) => {
    dispatch(ActionCreator.changeCity(city));
    dispatch(ActionCreator.setOffers(offers));
  }
});

export {App};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(App);
