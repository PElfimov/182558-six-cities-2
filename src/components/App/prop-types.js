import PropTypes from 'prop-types';

export default {
  offers: PropTypes.arrayOf(PropTypes.exact({
    id: PropTypes.string,
    isPremium: PropTypes.bool,
    cost: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    type: PropTypes.oneOf([`Private room`, `Apartment`]),
    coordinates: PropTypes.array
  })
  ).isRequired
};
