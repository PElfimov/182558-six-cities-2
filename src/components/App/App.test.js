import React from 'react';
import renderer from 'react-test-renderer';
import App from './app';

const offers = [
  {
    isPremium: true,
    cost: 120,
    name: `Beautiful & luxurious apartment at great location`,
    rating: 55,
    type: `Apartment`
  },
  {
    isPremium: false,
    cost: 132,
    name: `Wood and stone place`,
    rating: 80,
    type: `Apartment`
  },
  {
    isPremium: true,
    cost: 120,
    name: `Canal View Prinsengracht`,
    rating: 55,
    type: `Apartment`
  },
  {
    isPremium: true,
    cost: 80,
    name: `Nice, cozy, warm big bed apartment`,
    rating: 100,
    type: `Apartment`
  },
  {
    isPremium: true,
    cost: 120,
    name: `Wood and stone place`,
    rating: 88,
    type: `Private room`
  }
];

jest.mock(`../points-map/points-map`, () => jest.fn().mockReturnValue(null));

it(`App correctly renders after relaunch`, () => {
  const tree = renderer
    .create(<App offers={offers} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
