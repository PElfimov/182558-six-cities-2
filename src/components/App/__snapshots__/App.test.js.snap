// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn correctly renders after relaunch 1`] = `
<Switch>
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(WithCheckAuth)",
        "type": [Function],
      }
    }
    exact={true}
    path="/login"
  />
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(WithAuth)",
        "type": [Function],
      }
    }
    exact={true}
    path="/favorites"
  />
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(Offer)",
        "type": [Function],
      }
    }
    exact={true}
    path="/offer/:id"
  />
  <Route
    render={[Function]}
  />
</Switch>
`;
